docs/proposals/08-shoot-apiserver-via-sni.md

```diff
@@ -0,0 +1,396 @@
+         |          |
+         |          |                                   +------------------+
+         |          |                                   |                  |
+         |          |             proxy + lb            | Shoot API Server |
+         |          +-------------------+-------------->+                  |
+         |   get                                        | Cluster C        |
+         | configuration                                |                  |
+         |                                              +------------------+
+         |
+         v                                                  Configure
+      +--+--------------+         +---------------------+   via Istio
+      |                 |         |                     |   Custom Resources
+      |     Pilot       +-------->+   Seed API Server   +<------------------+
+      |                 |         |                     |
+      |                 |         |                     |
+      +-----------------+         +---------------------+
+```
+
+In this case the `internal` and `external` `DNSEntries` should be changed to the Network LoadBalancer's IP.
+
+### In-cluster communication to the apiserver
+
+In Kubernetes the API server is discoverable via the master service (`kubernetes` in `default` namespace). Today, this service can only be of type `ClusterIP` - making in-cluster communication to the API server impossible due to:
+
+- the client doesn't set the `Server Name` in the TLS handshake, if it attempts to talk to an IP address. In this case, the TLS handshake reaches the Envoy IngressGateway proxy, but it's rejected by it.
+- Kubernetes services can be of type `ExternalName`, but the master service is not supported by [kubelet](https://github.com/gardener/gardener/issues/1135#issuecomment-505317932).
+  - even if this is fixed in future Kubernetes versions, this problem still exists for older versions where this functionality is not available.
+
+Another issue occurs when the client tries to talk to the apiserver via the in-cluster DNS. For all Shoot API servers `kubernetes.default.svc.cluster.local` is the same and when a client tries to connect to that API server using that server name. This makes distinction between different in-cluster Shoot clients impossible by the Envoy IngressGateway.
+
+To mitigate this problem an additional proxy must be deployed on every single Node. It does not terminate TLS and sends the traffic to the correct Shoot API Server. This is achieved by:
+
+- the apiserver master service reconciler is stopped (`--endpoint-reconciler-type=none`).
+- the proxy runs in the host network of the `Node`.
+- the proxy has a sidecar container which:
+  - creates a dummy network interface and assigns the master service IP address to it.
+  - enables (with iptables) traffic to that IP address (kube-proxy disables it) and it removes connection tracking (conntrack) as the IP address is local to the `Node`.
```

Comment:
Two points here: 
- What do you mean by `kube-proxy` disables it. There is always a rule for service IP that gets dnated to endpoints? 
-  In this case we rely on IP tables, if eliminate IP-tables the solution needs to be re-worked.

Reply:
> What do you mean by kube-proxy disables it. There is always a rule for service IP that gets dnated to endpoints?

I'll reword it to `reject` as this is what `kube-proxy` does to services without endpoints:

```
A KUBE-SERVICES -d 10.105.243.238/32 -p tcp -m comment --comment "default/kubernetes-foo:https has no endpoints" -m tcp --dport 443 -j REJECT --reject-with icmp-port-unreachable
```

> In this case we rely on IP tables, if eliminate IP-tables the solution needs to be re-worked.

I've mentioned this in the future work.